Extension/src/Debugger/configurationProvider.ts

```diff
@@ -71,67 +80,149 @@ export class QuickPickConfigurationProvider implements vscode.DebugConfiguration
+            Telemetry.logDebuggerEvent(DebuggerEvent.debugPanel, { "debugType": "debug", "folderMode": folder ? "folder" : "singleMode", "cancelled": "true" });
+            return []; // User canceled it.
         }
-        if (selection.label.startsWith("cl.exe")) {
-            if (!process.env.DevEnvDir) {
-                vscode.window.showErrorMessage(localize("cl.exe.not.available", "{0} build and debug is only usable when VS Code is run from the Developer Command Prompt for VS.", "cl.exe"));
-                return [selection.configuration];
-            }
-        }
-        if (selection.label.indexOf(buildAndDebugActiveFileStr()) !== -1 && selection.configuration.preLaunchTask) {
-            try {
-                await cppBuildTaskProvider.checkBuildTaskExists(selection.configuration.preLaunchTask);
-                if (selection.configuration.miDebuggerPath) {
-                    if (!await util.checkFileExists(selection.configuration.miDebuggerPath)) {
-                        vscode.window.showErrorMessage(localize("miDebuggerPath.not.available", "miDebuggerPath does not exist: {0}. Has a debugger been installed?", selection.configuration.miDebuggerPath));
-                        throw new Error();
-                    }
-                }
-                await vscode.debug.startDebugging(folder, selection.configuration);
-                Telemetry.logDebuggerEvent("buildAndDebug", { "success": "true" });
-            } catch (e) {
-                Telemetry.logDebuggerEvent("buildAndDebug", { "success": "false" });
-            }
+        if (!this.isClAvailable(selection.label)) {
+            return [selection.configuration];
         }
+
         return [selection.configuration];
     }
 
+    /**
+	 * Error checks the provided 'config' without any variables substituted.
+	 */
     resolveDebugConfiguration(folder: vscode.WorkspaceFolder | undefined, config: vscode.DebugConfiguration, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration> {
-        return this.underlyingProvider.resolveDebugConfiguration ? this.underlyingProvider.resolveDebugConfiguration(folder, config, token) : undefined;
+        if (!config || !config.type) {
+            // When DebugConfigurationProviderTriggerKind is Dynamic, this function will be called with an empty config.
+            // Providing debug configs, and debugging should be called manually.
+            this.provideDebugConfigurations(folder).then(async configs => {
+                if (!configs || configs.length === 0) {
+                    Telemetry.logDebuggerEvent(DebuggerEvent.debugPanel, { "debugType": "debug", "folderMode": folder ? "folder" : "singleMode", "cancelled": "true" });
+                    return undefined; // aborts debugging silently
+                } else {
+                    await this.startDebugging(folder, configs[0], DebuggerEvent.debugPanel);
```

Comment:
It might be good to log telemetry or `console.assert` that `configs.length === 1`.  Currently the provider function returns only 1 result, but if that ever changes, we could end up doing the wrong thing here.  At the very least, add a comment here that there's only 1 config expected in the array.